_source/_posts/2022-01-12-webassembly-on-kubernetes-with-rust.md

```diff
@@ -0,0 +1,351 @@
+
+        thread::sleep(time::Duration::new(60, 0));
+    }
+}
+```
+
+The code is simple. It makes a GET request to the API and parses and prints the response every 60 seconds. Now you can build this into a Wasm binary using the below cargo command:
+
+```bash
+cargo build --release --target wasm32-wasi
+```
+
+That's it. You have successfully created a WebAssembly binary using Rust.
+
+## Run the workload in Kubernetes
+
+<!--
+Let's run the workload locally using wasmtime, a small JIT-style runtime for Wasm and WASI. Install it using the below command.
+
+```bash
+curl https://wasmtime.dev/install.sh -sSf | bash
+``` -->
+
+Before you can run the workload in Kubernetes, you need to push the binary to a registry that supports OCI artifacts. OCI complaint registries can be used for any OCI artifact, including Docker images, Wasm binaries, and so on. Docker Hub currently does not support OCI artifacts; hence you can use another registry like [GitHub Package Registry](https://github.com/features/packages), [Azure Container Registry](https://azure.microsoft.com/en-in/services/container-registry/) or [Google Artifact Registry](https://cloud.google.com/artifact-registry). I'll be using GitHub Package Registry as it's the simplest to get started, and most of you might already have a GitHub account.
+
+First, you need to log in to GitHub Package Registry using `docker login`. Create a [personal access token on GitHub](https://github.com/settings/tokens) and use that to log in to the registry.
+
+```bash
+export CR_PAT=<your-token>
+echo $CR_PAT | docker login ghcr.io -u <Your GitHub username> --password-stdin
+```
+
+Now you need to push your Wasm binary as an OCI artifact; for this, you can use the [wasm-to-oci](https://github.com/engineerd/wasm-to-oci) CLI. Use the below command to install it on your machine. I'm using the Linux version.
+
+```bash
+curl -LO https://github.com/engineerd/wasm-to-oci/releases/download/v0.1.2/linux-amd64-wasm-to-oci
+mv linux-amd64-wasm-to-oci ~/bin/wasm-to-oci # move to any location that is added to your PATH variable
+```
+
+Now you can push the binary you built earlier to the GitHub Package Registry. Run the below command from the `rust-wasm` folder.
+
+```bash
+wasm-to-oci push target/wasm32-wasi/release/rust-wasm.wasm ghcr.io/<your GitHub user>/rust-wasm:latest
+```
+
+You should see a successful message. Now check your GitHub packages page on your profile, and you should see the artifact listed.
+
+{% img blog/webassembly-on-kubernetes-with-rust/gchr_repo.png alt:"GitHub Packages Registry" width:"800" %}{: .center-image }
+
+By default, the artifact would be private, but you need to make it public so that you can access it from the Krustlet cluster. Click the package name and click the **Package settings** button, scroll down, click **Change visibility**, and change to **public**.
+
+You can check this by pulling the artifact using the below command:
+
+```bash
+ghcr.io/<your GitHub user>/rust-wasm:latest
```

Comment:
```suggestion
wasm-to-oci pull ghcr.io/<your GitHub user>/rust-wasm:latest
```

Reply:
Oops I think I forgot to finish the command